### PR TITLE
[Edit Mode] #627 Remove edit-related concerns from Browse controllers

### PR DESCRIPTION
**Pull request by akhenry**
Ok, so in order to un-block this ticket I am removing the dependency on regions. Actually it can be fairly simply achieved without resorting to regions with one additional representation ('view-object').

### Summary of Changes
* Removed all edit concerns from BrowseController and BrowseObjectController
* Removed most edit concerns from markup. The 'elements' section is still present in the browse inspector.  It might be desirable to refactor this out when the regions API has been decided on. I don't believe this is a major issue as it is though. By default it is hidden and will only get enable if object is in edit mode. In terms of the litmus test of disabling the edit bundle, the only ill-effect is a warning about two missing representations. Post #463 we may be able to neatly refactor this out via regions.
* Changed the handling of unload warnings. In new edit mode, navigation occurs through the navigation service. I have modified mct-unload to also capture navigation events via the navigation service. It does this via a listener on the NavigationService. This simplified navigation warnings (which previously tended to be triggered after a navigation change had taken place (via angular location change), requiring the need to manually reset the navigated object. This should reduce the number of edge cases around dirty navigation warnings
* Re-enabled disabled tests relating to browse controllers, and updated navigation service spec for new listeners

### Author Checklist

1. Changes address original issue? __Y__
2. Unit tests included and/or updated with changes? __Y__
3. Command line build passes? __Y__
4. Changes have been smoke-tested? __Y__

### CODE REVIEW

File: platform/commonUI/edit/src/representers/EditRepresenter.js [hunk 1]
```diff
@@ -111,11 +120,33 @@ define(
...
+            });
+
+            if (representedObject.getCapability('status').get('editing')){
+                setEditing();
+            }
         };
 
         // Respond to the destruction of the current representation.
         EditRepresenter.prototype.destroy = function destroy() {
             // Nothing to clean up
```
Comment (VWoeltjen) [inline platform/commonUI/edit/src/representers/EditRepresenter.js]: This comment looks out of date now

File: platform/representation/src/gestures/DropGesture.js [hunk 2]
```diff
@@ -165,16 +165,18 @@ define(
...
                     } else {
                         $q.when(action && action.perform()).then(function (result) {
                             //Don't go into edit mode for folders
                             if (domainObjectType!=='folder') {
-                                navigationService.setNavigation(editableDomainObject);
-                                editableDomainObject.getCapability('status').set('editing', true);
+                               // navigationService.setNavigation(editableDomainObject);
+                                //editableDomainObject.getCapability('status').set('editing', true);
+                                editableDomainObject.getCapability('action').perform('edit');
                             }
```
Comment (VWoeltjen) [inline platform/representation/src/gestures/DropGesture.js]: Like the changes here to use actions. There are some commented-out lines that should just be removed.

File: platform/representation/src/gestures/DropGesture.js [hunk 3]
```diff
@@ -165,16 +165,18 @@ define(
                     if (shouldCreateVirtualPanel(domainObject, selectedObject)){
```
Comment (VWoeltjen) [inline platform/representation/src/gestures/DropGesture.js]: Some of this logic is still rather NEM-specific. I think this stems from a shortcoming in the gesture API, though, so don't think we necessarily need to address this here. What we really want is to introduce additional logic between the physical gesture (drag-and-drop) and the consequence (choose between create virtual panel, etc). Right now, gestures and their consequences are not clearly separated, though, so this isn't really feasible. Want to consider this in the context of #463

**Comment by VWoeltjen** [comment]
I like the approach of making `setNavigation` an operation that may fail (returning true/false). The way this integrates with `NavigationService` (adding events to the event cycle) and `MCTBeforeUnload` (which previously was unconcerned with the information model) seems like it's mixing concerns

Could we address this as a `navigation` policy instead? e.g. `policyService.allow('navigation', newObject, priorObject)`. This could be wired into `NavigationService` as a decorator (meaning `NavigationService` itself can remain mostly unchanged, aside from returning `true` and having some JSDoc to reflect that). A policy instance could then be implemented using the same code from `EditObjectController.getUnloadWarning`; `getUnloadWarning` could in turn be rewritten to check that same policy (for cases where browser action triggers the `mct-before-unload`). I think this would result in fewer changes to existing classes, keeping units small and focused.

A few other comments in-line. In general, though, I think this is highly successful at reducing the extent to which edit-related concerns appear in non-edit-related code.

File: platform/representation/src/gestures/DropGesture.js (hunk 3 above)
Comment (akhenry) [inline platform/representation/src/gestures/DropGesture.js]: Yeah I totally agree. Actually the intention here was not to do the dropGesture refactoring just yet. I intend to defer it for #636. I did some limited refactoring that was necessitated by the changes to navigation. I will remove the commended code though, that was just an oversight.

**Comment by akhenry** [comment]
> Could we address this as a navigation policy instead? e.g. policyService.allow('navigation', newObject, priorObject)

This actually makes a lot of sense. I'll refactor it and see how it looks. Thanks

**Comment by akhenry** [comment]
There's a challenge here in showing an appropriate error message to the user when they try to navigate away from an object with a dirty model.

Switching to a policy means that the setNavigation call can fail and return false, but it's difficult to provide feedback to the user about why it failed. All that is known is that some check in the policy chain failed. Previously the dirty check was done in EditObjectController, via MctBeforeUnload, so it was easy to trigger a context aware user prompt. A policy isn't an appropriate place to show a user prompt.  

The result of setNavigation could be used from the calling point (`browseController`) to determine whether the navigation event was successful and to inform the user, but it won't really know why it failed. This is really a shortcoming of the policy API generally in that it doesn't easily support providing feedback about why an individual policy check failed. An example of where I have seen this issue addressed in other APIs is EMF's validation support where an error object is passed into each chained validation method, which can be used to accumulate error messages etc. I'm not suggesting that EMF should be used as any sort of exemplar though :)

For now, because we have only one 'navigation' policy it's possible to put the prompt in the controller (though I would move it from `BrowseController` to `BrowseObjectController` and `EditObjectController`). It's not very future proof though, any future navigation policy failures will give the same prompt, regardless of failure reason.

One thought that I had was policies could return an error string on failure. This is dangerous with the current binary policy model though, because tests for truthiness will pass, which might cause unexpected behavior.

Do you have any thoughts on the best place to put this user prompt?

**Comment by akhenry** [comment]
On second thought, maybe a user prompt from a policy check isn't the worst thing in the world. Essentially it's saying "The code can't make this policy decision without some user input". The user input becomes input into a policy decision.

It sets a precedent though I think?

**Comment by VWoeltjen** [comment]
There _is_ actually a mechanism in the policy API for returning reasons for failure, albeit not an especially elegant or obvious one. You can [pass a callback parameter](http://nasa.github.io/openmctweb/guide/#policy-service) to `policyService.allow`, and individual policies can [declare a failure message](http://nasa.github.io/openmctweb/guide/#policies-category).

It's not pretty, but we could also display a synchronous browser confirmation dialog _from_ the policy in this case, and decide whether or not to change navigation based on that. (This is consistent with the style of dialog shown for actual browser navigation changes - so there's that, at least.)

Looking forward to refactored Policy (in the context of #463), we could go asynchronous, such that usage is like:

```js
somePolicy.allow(something, someContext)
    .then(doSomething)
    .catch(handleError);
```

...which would have the desired bubble-up behavior. Such a change would be out of scope for this issue, but think it's useful to consider options in the context of use (esp. as we encounter shortcomings.)

The option I'd like to avoid is having a PolicyResult typedef/interface that PolicyService returns. We used that in the desktop client, but cases where we looked at anything other than the boolean were _exceptionally_ rare. (DRY usually causes you to stick policy down at lower-level calls, so by the time you're evaluating a policy decision you are often not in a good position to do anything with that message; that's where a bubble-up approach would be nice.)

**Comment by akhenry** [comment]
> There is actually a mechanism in the policy API for returning reasons for failure, albeit not an especially elegant or obvious one. You can pass a callback parameter to policyService.allow, and individual policies can declare a failure message.

Ah, my apologies, I completely missed this. I was only considering the Policy interface, and not the PolicyService.

That said, I've convinced myself now that the policy itself is actually the correct place to launch the confirmation dialog :) It's simply a policy decision that has a human in loop.

I could use the callback, but it still only gets me halfway there. The policyService is called from the navigation service decorator. It's still not a great place to have UI concerns such as user prompts, so it would still require some refactoring further upstream (in the controllers) which on analysis might be a little messy. 

Putting it in the policy itself is much cleaner than the alternatives I think.

**Comment by VWoeltjen** [comment]
> On second thought, maybe a user prompt from a policy check isn't the worst thing in the world. Essentially it's saying "The code can't make this policy decision without some user input". The user input becomes input into a policy decision.
>
> It sets a precedent though I think?

Yeah, I have mixed feelings about this as a precedent. It's could be very convenient/DRY to model user confirmations with policy. (Esp. if we switched to an asynchronous policy model, allowing us to use our own look-and-feel for such dialogs.)  On the other hand, it also seems like it mixes concerns - policy belongs in the information model, which shouldn't be making UI calls.

I think there are three future scenarios here, as the software evolves:

* We see this dialog when and only when we expect to see it. Great, it works; the "mixing concerns" problem was strictly academic.
* We see this dialog when in other circumstances, but it makes sense. Great - we've validated user confirmation as a policy concern!
* We see this dialog when it doesn't make sense. Bad - clearly we shouldn't have embedded UI interactions at this level.

Subjectively, I think the first option is the most likely, and (in the absence of other information) the latter two options seem equally likely. So I'd say go for it and we'll fix it if it breaks.

**Comment by akhenry** [comment]
Promises sounds good going forward. As well as feedback, it allows for asynchronous policy decisions (Though I would hope they would be used sparingly. Don't want lots of long chains of asynchronous policies being executed regularly)

**Comment by akhenry** [comment]
> So I'd say go for it and we'll fix it if it breaks.

It broke. :) It's incompatible with the before unload behavior.

I've pushed a changeset that does the user prompting in the controller. It's done from the BrowseController but is agnostic of failure causes, using the callback you pointed out, and policy failure message support. This actually makes the decorator redundant, so I've not pushed it.

I will defer fixing the tests until we've settled on an approach for this. Let me know if this looks ok.

**Comment by larkin** [comment]
I think we shouldn't mix user interactions with the policy service-- best to keep a strong separation of concerns.  I also think we should stick with a synchronous check for navigation; we don't need promises here, and many debugging messes are caused by methods that are sometimes asynchronous and other times synchronous. If you need to support asynchronous navigation checks, then it is best to initially prevent the navigation (but store the destination), show a dialog, and if the users choose to navigate, retry the navigation and auto-allow it.  This makes it easy to reason about the navigation state of the application.

The navigation service itself is a bit odd, we are working away from a lot of standard navigation concepts and the logical flow of navigating is obscure.  

Policy is an extension point, is it really correct to add a new policy category for navigation just to handle this specific portion of edit mode -- which has no demonstrated need for extensibility outside of the the core platform?

As an issue for another ticket, I think we need to rethink our route and navigation management.  Common solutions allow developers to declare routes that can be navigated to (register them with a router), and to have one method for navigating to a route (say, href changes).  The application router may have some lifecycle events-- e.g. `before:unload`, `after:load`, `load` as well as some delegated methods, such as "shouldNavigateToRoute" which could be defined by individual routes in order to implement their own logic for preventing navigation.

**Comment by VWoeltjen** [comment]
> I've pushed a changeset that does the user prompting in the controller. It's done from the BrowseController but is agnostic of failure causes, using the callback you pointed out, and policy failure message support. This actually makes the decorator redundant, so I've not pushed it.

I like moving the UI interaction into the UI layer, but BrowseController doesn't seem like an appropriate initiator. (What about navigation changes triggered from places other than the tree, for instance?)

Another option would be to address this as a decoration of the `navigate` action, added by Edit mode; BrowseController could in turn simplify `setNavigation` to `domainObject.getCapability('action').perform('navigate')`. This has a beneficial side effect of decreasing BrowseController's responsibilities slightly (consistent with the goals of the original issue). BrowseController does need to reset the tree if the user cancels; I would model this by catching rejection of the promise returned by the action.

The decorator's responsibility would be to check for unsaved changes, and prompt for user confirmation before delegating to the decorated action. (The promise returned by `dialogService.getUserInput` is rejected on user cancellation, so this bubbles nicely.) This could either take the place of the `navigate` policy (to @larkin's point, we don't have other use cases for this extension point) or simply use it. I'm leaning toward the former at this point; using policy would imply that _all_ such `navigate` policy decisions may be circumvented by the user, which is a big assumption.

Enforcing rules like this (expressed as policy or not) at the action level as opposed to the service level is also consistent with what we've done elsewhere (e.g. move/copy/link)

**Comment by larkin** [comment]
I'd push away from using actions for navigation. True, the browse controller is the wrong location for changing routing behavior, but it's done that way because the navigation system is currently intermingled with the browse controller / edit controller / action model and is a total mess.  

Let's not intermingle further, let's separate concerns.  We need to extricate navigation from actions and the browse controller and move to a more comprehensive routing and navigation system.  Then, the edit controller (or route, or whatever) can temporarily modify global navigation functionality as needed, without having to reach deep into other corners of the application.

File: platform/representation/src/gestures/DropGesture.js (hunk 3 above)
Comment (larkin) [inline platform/representation/src/gestures/DropGesture.js]: Can we pull out virtual panel creation from the drop gesture for now, seeing as they are already a temporary idea?  This behavior is a nice for functional demos but complicates integrations due to how it is hardcoded.  Would also simplify #636.

**Comment by akhenry** [comment]
@larkin so your vote would be to leave it in the BrowseController for now, with a view to refactoring navigation out of the BrowseController in future?
@VWoeltjen A point of clarification. I am assuming that this would be a decorator of the ActionProvider which will decorate NavigationAction before it's returned, rather than being defined as a decorator of the Action itself (which I didn't think could be decorated)?

**Comment by larkin** [comment]
@akhenry Yes, my vote would be to leave in browse controller, put in a book mark to factor out navigation and routing.

File: platform/representation/src/gestures/DropGesture.js (hunk 3 above)
Comment (akhenry) [inline platform/representation/src/gestures/DropGesture.js]: I would not be opposed to this, especially given that it doesn't work in VISTA and actually just complicates things there. It was just a prototype after all. 

I will leave it enabled in a branch if we ever want to demo it in WARP.

Comment (akhenry) [inline platform/representation/src/gestures/DropGesture.js]: @VWoeltjen Do we have consensus on this?

Comment (VWoeltjen) [inline platform/representation/src/gestures/DropGesture.js]: Sounds good to me

**Comment by VWoeltjen** [comment]
That works for me as well - the rationale makes sense.

> @VWoeltjen A point of clarification. I am assuming that this would be a decorator of the ActionProvider which will decorate NavigationAction before it's returned, rather than being defined as a decorator of the Action itself (which I didn't think could be decorated)?

A moot point now, I think, but yes, this would be necessary. This is a shortcoming of the current extension mechanism that I believe is addressed in #705; I'll add an example to that document to illustrate this.

**Comment by VWoeltjen** [comment]
Added the [example referenced above](https://github.com/nasa/openmctweb/blob/open462/docs/src/design/proposals/Registration.md#extending-extensions)

**Comment by akhenry** [comment]
### Changes
* NavigationService and MCTBeforeUnload reverted to original state
* Added EditNavigationPolicy that is used (indirectly, via PolicyService) by BrowseController and EditObjectController for confirming navigation and browse location change events respectively.
* Removed "virtual panels" code from DropGesture. Drag-drop auto-creation of panels no longer supported.
* Fixed failing tests
* Rebased over master

**Comment by VWoeltjen** [comment]
### Reviewer Checklist

1. Changes appear to address issue? Y
2. Appropriate unit tests included? Y
3. Code style and in-line documentation are appropriate? Y
4. Commit messages meet standards? Y
